primitives/src/math/fixed.rs

```diff
@@ -0,0 +1,824 @@
+}
+
+/// Converts a fixed point decimal number into another type.
+pub trait IntoFixedFromDecimal<F> {
+    /// Converts a fixed point decimal number into another type.
+    fn to_fixed_from_fixed_decimal(self, places: u8) -> Result<F, ParseFixedError>;
+}
+
+/// Converts a type into a fixed point decimal number.
+pub trait FromFixedToDecimal<F>
+where
+    Self: Sized + TryFrom<u128>,
+{
+    /// Craft a fixed point decimal number from another type.
+    fn from_fixed_to_fixed_decimal(fixed: F, places: u8) -> Result<Self, &'static str>;
+}
+
+/// Converts a type into a fixed point decimal number.
+pub trait IntoFixedDecimal<N: TryFrom<u128>> {
+    /// Converts a type into a fixed point decimal number.
+    fn to_fixed_decimal(self, places: u8) -> Result<N, &'static str>;
+}
+
+impl<F: Fixed, N: Into<u128>> FromFixedDecimal<N> for F {
+    /// Craft a `Fixed` type from a fixed point decimal number of type `N`
+    fn from_fixed_decimal(decimal: N, places: u8) -> Result<Self, ParseFixedError> {
+        let decimal_u128 = decimal.into();
+        let mut decimal_string = decimal_u128.to_string();
+
+        if decimal_string.len() <= places as usize {
+            // This can never underflow (places >= len). Saturating subtraction to satisfy clippy.
+            decimal_string = "0.".to_owned()
+                + &"0".repeat((places as usize).saturating_sub(decimal_string.len()))
+                + &decimal_string;
+        } else {
+            // This can never underflow (len >= places). Saturating subtraction to satisfy clippy.
+            decimal_string.insert(decimal_string.len().saturating_sub(places as usize), '.');
+        }
+
+        F::from_str(&decimal_string)
+    }
+}
+
+impl<F, N> IntoFixedFromDecimal<F> for N
+where
+    F: Fixed + FromFixedDecimal<Self>,
+    N: Into<u128>,
+{
+    /// Converts a fixed point decimal number into `Fixed` type (e.g. `Balance` -> `Fixed`).
+    fn to_fixed_from_fixed_decimal(self, places: u8) -> Result<F, ParseFixedError> {
+        F::from_fixed_decimal(self, places)
+    }
+}
+
+impl<F: Fixed + ToString, N: TryFrom<u128>> FromFixedToDecimal<F> for N {
+    fn from_fixed_to_fixed_decimal(fixed: F, decimals: u8) -> Result<N, &'static str> {
+        let decimals_usize = decimals as usize;
+        let s = fixed.to_string();
+        let mut parts: Vec<&str> = s.split('.').collect();
+        // If there's no fractional part, then `fixed` was an integer.
+        if parts.len() != 2 {
+            parts.push("0");
+        }
+
+        let (int_part, frac_part) = (parts[0], parts[1]);
+        let mut increment = false;
+
+        let new_frac_part = if frac_part.len() < decimals_usize {
+            format!("{}{}", frac_part, "0".repeat(decimals_usize - frac_part.len()))
+        } else {
+            // Adding rounding behavior
+            let round_digit = frac_part.chars().nth(decimals_usize);
+            match round_digit {
+                Some(d) if d >= '5' => increment = true,
```

Comment:
Admittedly, this isn't ideal. I just didn't feel like adding another conversion. This will fail spectacularly if the device doesn't properly implement the unicode standard. Probably something we should fix in a future version.

Reply:
Can you add an issue for that?

Reply:
https://github.com/zeitgeistpm/zeitgeist/issues/1131